src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,143 @@
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode;
+        if (first == null) {
+            newNode = new Node<>(null, value, null);
+            first = newNode;
+        } else {
+            newNode = new Node<>(last, value, null);
+            last.next = newNode;
+        }
+        last = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        Node<T> newNode;
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            newNode = new Node<>(null, value, first);
+            first.prev = newNode;
+            first = newNode;
+            size++;
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index);
+            newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = newNode;
+            nodeByIndex.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> deleteNode = getNodeByIndex(index);
+        if (isEmpty()) {
+            throw new RuntimeException("List is Empty!");
+        } else if (size == 1) {
+            first = null;
+            last = null;
+        } else if (deleteNode == first) {
+            first = deleteNode.next;
+            first.prev = null;
+        } else if (deleteNode == last) {
+            last = deleteNode.prev;
+            last.next = null;
+        } else {
+            deleteNode.prev.next = deleteNode.next;
+            deleteNode.next.prev = deleteNode.prev;
+        }
+        size--;
+        return deleteNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        for (int i = 0; i < size; i++) {
+            if (current.item == object
+                    || (current.item != null) && (current.item).equals(object)) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index, size);
```

Comment:
Dont use 'checkIndex' here, because you use this method before you call 'getNodeByIndex'

Reply:
I replaced calling method checkIndex() in all another times, when I call getNodeByIndex now))